docs/reference/configuration.md
---
mapped_pages:
  - https://www.elastic.co/guide/en/apm/agent/ruby/current/configuration.html
applies_to:
  stack:
  serverless:
    observability:
  product:
    apm_agent_ruby: ga
---

# Configuration [configuration]

There are several ways to configure how Elastic APM behaves. We recommend using a `config/elastic_apm.yml` file:

```yaml
server_url: 'http://localhost:8200'
secret_token: <%= ENV["VERY_SECRET_TOKEN"] %>
```

Some options can be set with `ENV` variables. When using this method, strings are split by comma, e.g., `ELASTIC_APM_SANITIZE_FIELD_NAMES="a,b" # => [/a/, /b/]`.


## Configuration precedence [configuration-precedence]

Options are applied in the following order (last one wins):

1. Defaults
2. Arguments to `ElasticAPM.start` / `Config.new`
3. Config file, e.g., `config/elastic_apm.yml`
4. Environment variables
5. [Central configuration](docs-content://solutions/observability/apps/apm-agent-central-configuration.md) (supported options are marked with [![dynamic config](images/dynamic-config.svg "") ](#dynamic-configuration))


## Dynamic configuration [dynamic-configuration]

Configuration options marked with the ![dynamic config](images/dynamic-config.svg "") badge can be changed at runtime when set from a supported source.

The Agent supports [Central configuration](docs-content://solutions/observability/apps/apm-agent-central-configuration.md), which allows you to fine-tune certain configurations via the APM app. This feature is enabled in the Agent by default, with [`central_config`](#config-central-config).


## Ruby on Rails [_ruby_on_rails]

When using Rails, it’s possible to specify options inside `config/application.rb`:

```ruby
# config/application.rb
config.elastic_apm.service_name = 'MyApp'
```


## Sinatra and Rack [_sinatra_and_rack]

When using Sinatra and Rack, you can configure when starting the agent:

```ruby
# config.ru or similar
ElasticAPM.start(
  app: MyApp,
  service_name: 'SomeOtherName'
)
```

Alternatively, use the `ElasticAPM::Sinatra.start` API:

```ruby
# config.ru or similar
ElasticAPM::Sinatra.start(
  MyApp,
  service_name: 'SomeOtherName'
)
```

See [Getting started with Rack](getting-started-rack.md).


## Grape and Rack [_grape_and_rack]

When using Grape and Rack (without Rails), configure when starting the agent:

```ruby
# config.ru or similar
ElasticAPM::Grape.start(
  MyApp,
  service_name: 'SomeOtherName'
)
```

See [Getting started with Rack](getting-started-rack.md).


## Options [_options]


### `config_file` [config-config-file]

| Environment | `Config` key | Default |
| --- | --- | --- |
| `ELASTIC_APM_CONFIG_FILE` | `config_file` | `config/elastic_apm.yml` |

The path to the configuration YAML-file. Elastic APM will load config options from this if the file exists. The file will be evaluated as ERB, so you can include `ENV` variables like in your `database.yml`, eg:

```ruby
secret_token: <%= ENV['VERY_SECRET_TOKEN'] %>
```


### `server_url` [config-server-url]

| Environment | `Config` key | Default |
| --- | --- | --- |
| `ELASTIC_APM_SERVER_URL` | `server_url` | `'http://localhost:8200'` |

The URL for your APM Server. The URL must be fully qualified, including protocol (`http` or `https`) and port.


### `secret_token` [config-secret-token]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_SECRET_TOKEN` | `secret_token` | `nil` | A random string |

This string is used to ensure that only your agents can send data to your APM server. Both the agents and the APM server have to be configured with the same secret token. Here’s an example that generates a secure secret token:

```bash
ruby -r securerandom -e 'print SecureRandom.uuid'
```

::::{warning}
Secret tokens only provide any real security if your APM server uses TLS.
::::



### `api_key` [config-api-key]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_API_KEY` | `api_key` | `nil` | A base64-encoded string |

This base64-encoded string is used to ensure that only your agents can send data to your APM server. The API key must be created using the [APM server command-line tool](docs-content://solutions/observability/apps/api-keys.md).

::::{warning}
API keys only provide any real security if your APM server uses TLS.
::::



### `api_buffer_size` [config-api-buffer-size]

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_API_BUFFER_SIZE` | `api_buffer_size` | `256` |

The maximum amount of objects kept in queue before sending to APM Server.

If you hit the limit, consider increasing the agent’s [worker pool size](#config-pool-size). If you don’t, the agent may have trouble connecting to APM Server. The [logs](#config-log-path) should tell you what went wrong.


### `api_request_size` [config-api-request-size]

[![dynamic config](images/dynamic-config.svg "") ](#dynamic-configuration)

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_API_REQUEST_SIZE` | `api_request_size` | `"750kb"` |

The maximum amount of bytes sent over one request to APM Server. The agent will open a new request when this limit is reached.

This must be provided in **[size format](#config-format-size)**.


### `api_request_time` [config-api-request-time]

[![dynamic config](images/dynamic-config.svg "") ](#dynamic-configuration)

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_API_REQUEST_TIME` | `api_request_time` | `"10s"` |

The maximum duration of a single streaming request to APM Server before opening a new one.

The APM Server has its own limit of 30 seconds before it will close requests. This must be provided in **[duration format](#config-format-duration)**.


### `breakdown-metrics` [config-breakdown-metrics]

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_BREAKDOWN_METRICS` | `breakdown_metrics` | `true` |

Enable or disable the tracking and collection of breakdown metrics. Setting this to `False` disables the tracking of breakdown metrics, which can reduce the overhead of the agent.

::::{note}
This feature requires APM Server and Kibana >= 7.3.
::::



### `capture_body` [config-capture-body]

[![dynamic config](images/dynamic-config.svg "") ](#dynamic-configuration)

|     |     |     |     |
| --- | --- | --- | --- |
| Environment | `Config` key | Default | Example |
| `ELASTIC_APM_CAPTURE_BODY` | `capture_body` | `"off"` | `"all"` |

The Ruby agent can optionally capture the request body (e.g. `POST` variables or JSON data) for transactions that are HTTP requests.

Possible values: `"errors"`, `"transactions"`, `"all"`, `"off"`.

If the request has a body and this setting is disabled, the body will be shown as `[SKIPPED]`.

::::{warning}
Request bodies often contain sensitive values like passwords and credit card numbers. We try to strip sensitive looking data from form bodies but don’t touch text bodies like JSON. If your service handles data like this, we advise to only enable this feature with care.
::::



### `capture_headers` [config-capture-headers]

[![dynamic config](images/dynamic-config.svg "") ](#dynamic-configuration)

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_CAPTURE_HEADERS` | `capture_headers` | `true` |

This indicates whether or not to attach the request headers to transactions and errors.


### `capture_elasticsearch_queries` [config-capture-elasticsearch-queries]

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_CAPTURE_ELASTICSEARCH_QUERIES` | `capture_elasticsearch_queries` | `false` |

This indicates whether or not to capture the body from requests in Elasticsearch.


### `capture_env` [config-capture-env]

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_CAPTURE_ENV` | `capture_env` | `true` |

This indicates whether or not to attach `ENV` from Rack to transactions and errors.


### `central_config` [config-central-config]

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_CENTRAL_CONFIG` | `central_config` | `true` |

This enables [APM Agent Configuration via Kibana](docs-content://solutions/observability/apps/apm-agent-central-configuration.md). If set to `true`, the client will poll the APM Server regularly for new agent configuration.

Usually APM Server determines how often to poll, but if not, set the default interval is 5 minutes.

::::{note}
This feature requires APM Server v7.3 or later. See [APM Agent central configuration](docs-content://solutions/observability/apps/apm-agent-central-configuration.md) for more information.
::::



### `cloud_provider` [config-cloud-provider]

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_CLOUD_PROVIDER` | `cloud_provider` | `"auto"` |

Specify the cloud provider for metadata collection. This defaults to `"auto"`, which means the agent uses trial and error to collect metadata from all supported cloud providers.

Valid options are `"auto"`, `"aws"`, `"gcp"`, `"azure"`, and `"none"`. If set to `"none"`, no cloud metadata will be collected. If set to any of `"aws"`, `"gcp"`, or `"azure"`, attempts to collect metadata will only be performed from the chosen provider.


### `disable_metrics` [config-disable_metrics]

|     |     |     |     |
| --- | --- | --- | --- |
| Environment | `Config` key | Default | Example |
| `ELASTIC_APM_DISABLE_METRICS` | `disable_metrics` | [] | `"*.cpu.*,system.memory.total"` |

A comma-separated list of dotted metrics names that should not be sent to the APM Server. You can use `*` to match multiple metrics. Matching is not case sensitive.


### `disable_send` [config-disable-send]

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_DISABLE_SEND` | `disable_send` | `false` |

This disables sending payloads to APM Server.


### `disable_start_message` [config-disable-start-message]

|     |     |     |
| --- | --- | --- |
| Environment | `Config` key | Default |
| `ELASTIC_APM_DISABLE_START_MESSAGE` | `disable_start_message` | `false` |

This disables the agents startup message announcing itself.


### `disable_instrumentations` [config-disabled-instrumentations]

| Environment | `Config` key | Default |
| --- | --- | --- |
| `ELASTIC_APM_DISABLE_INSTRUMENTATIONS` | `disable_instrumentations` | `['json']` |

Elastic APM automatically instruments select third-party libraries. Use this option to disable any of these.

Get an array of enabled instrumentations with `ElasticAPM.agent.config.enabled_instrumentations`.


### `enabled` [config-enabled]

| Environment | `Config` key | Default |
| --- | --- | --- |
| `ELASTIC_APM_ENABLED` | `enabled` | `true` |

Indicates whether or not to start the agent. If `enabled` is `false`, `ElasticAPM.start` will do nothing and all calls to the root API will return `nil`.


### `environment` [config-environment]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_ENVIRONMENT` | `environment` | From `ENV` | `"production"` |

The name of the environment this service is deployed in, e.g. "production" or "staging".

Environments allow you to easily filter data on a global level in the APM app. Be consistent when naming environments across agents. See [environment selector](docs-content://solutions/observability/apps/filter-application-data.md#apm-filter-your-data-service-environment-filter) in the APM app for more information.

Defaults to `ENV['RAILS_ENV'] || ENV['RACK_ENV']`.

::::{note}
This feature is fully supported in the APM app in Kibana versions >= 7.2. You must use the query bar to filter for a specific environment in versions prior to 7.2.
::::



### `filter_exception_types` [config-filter-exception-types]

|     |     |     |     |
| --- | --- | --- | --- |
| Environment | `Config` key | Default | Example |
| N/A | `filter_exception_types` | `[]` | `[MyApp::Errors::IgnoredError]` |

Use this to filter error tracking for specific error constants.


### `framework_name` [config-framework-name]

| Environment | `Config` key | Default |
| --- | --- | --- |
| `ELASTIC_APM_FRAMEWORK_NAME` | `framework_name` | Depending on framework |

The name of the used framework. For Rails or Sinatra, this defaults to `Ruby on Rails` and `Sinatra` respectively, otherwise defaults to `nil`.


### `framework_version` [config-framework-version]

| Environment | `Config` key | Default |
| --- | --- | --- |
| `ELASTIC_APM_FRAMEWORK_VERSION` | `framework_version` | Depending on framework |

The version number of the used framework. For Ruby on Rails and Sinatra, this defaults to the used version of the framework, otherwise, the default is `nil`.


### `global_labels` [config-global-labels]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_GLOBAL_LABELS` | `global_labels` | `nil` | `dept=engineering,rack=number8` |

Labels added to all events, with the format key=value[,key=value[,… ]].

::::{note}
This option requires APM Server 7.2 or greater, and will have no effect when using older server versions.
::::



### `hostname` [config-hostname]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_HOSTNAME` | `hostname` | `hostname` | `app-server01.example.com` |

The host name to use when sending error and transaction data to the APM server.


### `ignore_url_patterns` [config-custom-ignore-url-patterns]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_IGNORE_URL_PATTERNS` | `ignore_url_patterns` | `[]` | `['^/ping', %r{^/admin}]` |

Use this option to ignore certain URL patterns such as healthchecks or admin sections.

*Ignoring* in this context means *don’t wrap in a [Transaction](api-reference.md#api-transaction)*. Errors will still be reported.

Use a comma separated string when setting this option via `ENV`. Eg. `ELASTIC_APM_IGNORE_URL_PATTERNS="a,b" # => [/a/, /b/]`


### `instrument` [config-instrument]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_INSTRUMENT` | `instrument` | `true` | `0` |

Use this option to ignore certain URL patterns such as healthchecks or admin sections.


### `instrumented_rake_tasks` [config-instrumented-rake-tasks]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_INSTRUMENTED_RAKE_TASKS` | `instrumented_rake_tasks` | `[]` | `['my_task']` |

Elastic APM can instrument your Rake tasks. This is an opt-in field, as they are used are for a multitude of things.


### `log_ecs_reformatting` [config-log-ecs-formatting]

| Environment | `Config` key | Default |
| --- | --- | --- |
| `ELASTIC_APM_LOG_ECS_REFORMATTING` | `log_ecs_reformatting` | `off` |

This is an experimental option that configures the agent to use the logger from the `ecs-logging` gem. The two valid options are `off` and `override`.

Setting this option to `override` will set the agent logger to a `EcsLogging::Logger` instance and all logged output will be in ECS-compatible json.

The `ecs-logging` gem must be installed before the agent is started. If `log_ecs_reformatting` is set to `override`, the agent will attempt to require the gem and if it cannot be loaded, it will fall back to using the standard Ruby `::Logger` and log the load error.

Note that if you’re using Rails, the agent will ignore this option and use the Rails logger. If you want to use a `EcsLogging::Logger` when using Rails, set the agent’s logger config option explicitly to a `EcsLogging::Logger` instance.


### `log_level` [config-log-level]

[![dynamic config](images/dynamic-config.svg "") ](#dynamic-configuration)

| Environment | `Config` key | Default |
| --- | --- | --- |
| `ELASTIC_APM_LOG_LEVEL` | `log_level` | `Logger::INFO # => 1` |

By default Elastic APM logs to `stdout` or uses `Rails.log` when used with Rails.


### `log_path` [config-log-path]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_LOG_PATH` | `log_path` | `nil` | `log/elastic_apm.log` |

A path to a log file.

By default Elastic APM logs to `stdout` or uses `Rails.log` when used with Rails.

If `log_path` is specified, this will override `Rails.log` to point to that path instead.

This should support both absolute and relative paths. Please be sure the directory exists.


### `logger` [config-logger]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| N/A | `logger` | Depends | `Logger.new('path/to_file.log')` |

By default Elastic APM logs to `stdout` or uses `Rails.log` when used with Rails.

Use this to provide another logger. This is expected to have the same API as Ruby’s built-in `Logger`.


### `metrics_interval` [config-metrics-interval]

| Environment | `Config` key | Default |
| --- | --- | --- |
| `ELASTIC_APM_METRICS_INTERVAL` | `metrics_interval` | `'30s'` |

Specify the interval for reporting metrics to APM Server. The interval should be in seconds, or include a time suffix.

To disable metrics reporting, set the interval to `0`.


### `pool_size` [config-pool-size]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_POOL_SIZE` | `pool_size` | `1` | `2` |

Elastic APM uses a thread pool to send its data to APM Server.

This makes sure the agent doesn’t block the main thread any more than necessary.

If you have high load and get warnings about the buffer being full, increasing the worker pool size might help.


### `proxy_address` [config-proxy-address]

| Environment | `Config` key | Default | Example |
| --- | --- | --- | --- |
| `ELASTIC_APM_PROXY_ADDRESS` | `proxy_address` | `nil` | `"example.com"` |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

- Add `ElasticAPM::Sinatra.start` API. ([#566](https://github.com/elastic/apm-agent-ruby/pull/566))

#### Changed

- Log 404s from CentralConfig on debug level ([#553](https://github.com/elastic/apm-agent-ruby/pull/553))

#### Fixed

- Fix Central Config url path ([#578](https://github.com/elastic/apm-agent-ruby/pull/578))

## 3.0.0 (2019-10-08)

### Breaking Changes

The following changes are breaking, as they may change the way data is grouped or shown in Kibana or how your app
defines agent settings.

- Decrease stack_trace_limit to 50 ([#515](https://github.com/elastic/apm-agent-ruby/pull/515))
- Errors' `message` no longer include their `type` ([#323](https://github.com/elastic/apm-agent-ruby/pull/323/files))
- External request spans now have type `external.http.{library}` ([#514](https://github.com/elastic/apm-agent-ruby/pull/514))
- Durations are measured using monotonic time ([#550](https://github.com/elastic/apm-agent-ruby/pull/550))
- Rename set_tag to set_label ([#543](https://github.com/elastic/apm-agent-ruby/pull/543))
- Rename disabled_spies to disabled_instrumentations ([#539](https://github.com/elastic/apm-agent-ruby/pull/539))

## 2.12.0 (2019-10-01)

### Changed

- `disabled_spies` renamed to `disabled_instrumentations` with fallback ([#539](https://github.com/elastic/apm-agent-ruby/pull/539))
- Rename `set_tag` to `set_label` and deprecate `set_tag` ([#543](https://github.com/elastic/apm-agent-ruby/pull/543))
- Allow non-String label values ([#543](https://github.com/elastic/apm-agent-ruby/pull/543))

### Fixed

- Handles a case where stacktrace frames are empty ([#538](https://github.com/elastic/apm-agent-ruby/pull/538))

### Deprecated

- Deprecate `set_tag` ([#543](https://github.com/elastic/apm-agent-ruby/pull/543))

## 2.11.0 (2019-09-23)

### Added

- Add `Rails` module with `#start` method to run Rails setup explicitly ([#522](https://github.com/elastic/apm-agent-ruby/pull/522))
- Support for log/trace correlation ([#527](https://github.com/elastic/apm-agent-ruby/pull/527))

### Changed

- Split dot-separated `span.type` into `.type`, `.subtype` and `.action` (auto-upgrades dot style) ([#531](https://github.com/elastic/apm-agent-ruby/pull/531))

## 2.10.1

### Fixed

- Fixed loading options from a config file specified by `ELASTIC_APM_CONFIG_FILE` ([#518](https://github.com/elastic/apm-agent-ruby/pull/518))
- Fixed an issue with CentralConfig polling not starting ([#525](https://github.com/elastic/apm-agent-ruby/pull/525))

### Added

- Support for chained exceptions ([#488](https://github.com/elastic/apm-agent-ruby/pull/488))

## 2.10.0

### Added

- Add Ruby specific metrics ([#437](https://github.com/elastic/apm-agent-ruby/pull/437))
- Support for APM Agent Configuration via Kibana ([#464](https://github.com/elastic/apm-agent-ruby/pull/464))
- Change span name format and add command to context's db.statement for `MongoSpy` ([#488](https://github.com/elastic/apm-agent-ruby/pull/490))

### Changed

- `ElasticAPM.report` and `ElasticAPM.report_message` return the string ID of the generated `Error` objects ([#507](https://github.com/elastic/apm-agent-ruby/pull/507))

## 2.9.1 (2019-06-28)

### Fixed

- Use system CA certificate if none is specified ([#460](https://github.com/elastic/apm-agent-ruby/pull/460))

## 2.9.0 (2019-06-25)

### Security

- **NB:** If you are using a custom CA cert via `server_ca_cert`, versions of the agent prior to 2.9.0 may not have validated the certificate of APM Server correctly.

### Added

- Add `transaction.type` to errors ([#434](https://github.com/elastic/apm-agent-ruby/pull/434))
- Add cookies to `request.cookies` ([#448](https://github.com/elastic/apm-agent-ruby/pull/448))

### Fixed

- Fix support for older versions of Http.rb ([#438](https://github.com/elastic/apm-agent-ruby/pull/434))
- Strip `Cookie` and `Set-Cookie` from headers ([#448](https://github.com/elastic/apm-agent-ruby/pull/448))
- Fix disabling SSL verification ([#449](https://github.com/elastic/apm-agent-ruby/pull/449))

## 2.8.1 (2019-05-29)

### Fixed

- Database statements are properly truncated ([#431](https://github.com/elastic/apm-agent-ruby/pull/431))

## 2.8.0 (2019-05-20)

### Added

- The option `stack_trace_limit` ([#424](https://github.com/elastic/apm-agent-ruby/pull/424))

### Changed

- Use standardized User-Agent ([#419](https://github.com/elastic/apm-agent-ruby/pull/419))

### Fixed

- `error.culprit` is properly truncated to 1024 characters ([#418](https://github.com/elastic/apm-agent-ruby/pull/418))
- Force convert `transaction.context.response.status_code` to integer ([#423](https://github.com/elastic/apm-agent-ruby/pull/423))

## 2.7.0 (2019-05-07)

### Added

- Added `disable_start_message` for those wanting complete silence on boot ([#397](https://github.com/elastic/apm-agent-ruby/pull/397))

### Changed

- Attempt to strip secrets from `request.env` ([#410](https://github.com/elastic/apm-agent-ruby/pull/410))

### Fixed

- Rewritten most of the internal transport code, streaming events to APM Server ([#372](https://github.com/elastic/apm-agent-ruby/pull/372))
- Re-added `default_tags` ([#395](https://github.com/elastic/apm-agent-ruby/pull/395))
- A bug in the Faraday spy when disabling the Net::HTTP spy ([#396](https://github.com/elastic/apm-agent-ruby/pull/396))
- Fix disabling the ActionDispatch spy ([#399](https://github.com/elastic/apm-agent-ruby/pull/399))

## 2.6.1 (2019-03-28)

### Fixed

- Setting `config_file` via `ELASTIC_APM_CONFIG_FILE` ([#363(https://github.com/elastic/apm-agent-ruby/pull/363))
- Stopping the Metrics collector when it is disabled ([#357](https://github.com/elastic/apm-agent-ruby/pull/357))
- HTTP proxy settings can now be set by ENV variable ([#367](https://github.com/elastic/apm-agent-ruby/pull/367))


lib/elastic_apm/config.rb
# Licensed to Elasticsearch B.V. under one or more contributor
# license agreements. See the NOTICE file distributed with
# this work for additional information regarding copyright
# ownership. Elasticsearch B.V. licenses this file to you under
# the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

# frozen_string_literal: true

require 'elastic_apm/config/bytes'
require 'elastic_apm/config/duration'
require 'elastic_apm/config/log_level_map'
require 'elastic_apm/config/options'
require 'elastic_apm/config/round_float'
require 'elastic_apm/config/round_float_hash_value'
require 'elastic_apm/config/regexp_list'
require 'elastic_apm/config/wildcard_pattern_list'
require 'elastic_apm/deprecations'
require 'elastic_apm/config/server_info'

module ElasticAPM
  # @api private
  class Config
    extend Options
    extend Deprecations

    SANITIZE_FIELD_NAMES_DEFAULT =
      %w[password passwd pwd secret *key *token* *session* *credit* *card* *auth* set-cookie].freeze

    # rubocop:disable Layout/LineLength, Layout/ExtraSpacing
    option :config_file,                       type: :string, default: 'config/elastic_apm.yml'
    option :server_url,                        type: :url,    default: 'http://localhost:8200'
    option :secret_token,                      type: :string
    option :api_key,                           type: :string

    option :api_buffer_size,                   type: :int,    default: 256
    option :api_request_size,                  type: :bytes,  default: '750kb', converter: Bytes.new
    option :api_request_time,                  type: :float,  default: '10s',   converter: Duration.new
    option :breakdown_metrics,                 type: :bool,   default: true
    option :capture_body,                      type: :string, default: 'off'
    option :capture_headers,                   type: :bool,   default: true
    option :capture_elasticsearch_queries,     type: :bool,   default: false
    option :capture_env,                       type: :bool,   default: true
    option :central_config,                    type: :bool,   default: true
    option :cloud_provider,                    type: :string, default: 'auto'
    option :current_user_email_method,         type: :string, default: 'email'
    option :current_user_id_method,            type: :string, default: 'id'
    option :current_user_username_method,      type: :string, default: 'username'
    option :custom_key_filters,                type: :list,   default: [],      converter: RegexpList.new
    option :default_labels,                    type: :dict,   default: {}
    option :disable_metrics,                   type: :list,   default: [],      converter: WildcardPatternList.new
    option :disable_send,                      type: :bool,   default: false
    option :disable_start_message,             type: :bool,   default: false
    option :disable_instrumentations,          type: :list,   default: %w[json]
    option :disabled_spies,                    type: :list,   default: []
    option :enabled,                           type: :bool,   default: true
    option :environment,                       type: :string, default: ENV['RAILS_ENV'] || ENV['RACK_ENV']
    option :framework_name,                    type: :string
    option :framework_version,                 type: :string
    option :filter_exception_types,            type: :list,   default: []
    option :global_labels,                     type: :dict
    option :hostname,                          type: :string
    option :http_compression,                  type: :bool,   default: true
    option :ignore_url_patterns,               type: :list,   default: [],      converter: RegexpList.new
    option :instrument,                        type: :bool,   default: true
    option :instrumented_rake_tasks,           type: :list,   default: []
    option :log_ecs_reformatting,              type: :string, default: 'off'
    option :log_level,                         type: :int,    default: Logger::INFO, converter: LogLevelMap.new
    option :log_path,                          type: :string
    option :metrics_interval,                  type: :int,    default: '30s',   converter: Duration.new
    option :pool_size,                         type: :int,    default: 1
    option :proxy_address,                     type: :string
    option :proxy_headers,                     type: :dict
    option :proxy_password,                    type: :string
    option :proxy_port,                        type: :int
    option :proxy_username,                    type: :string
    option :recording,                         type: :bool,   default: true
    option :sanitize_field_names,              type: :list,   default: SANITIZE_FIELD_NAMES_DEFAULT, converter: WildcardPatternList.new
    option :server_ca_cert_file,               type: :string
    option :service_name,                      type: :string
    option :service_node_name,                 type: :string
    option :service_version,                   type: :string
    option :source_lines_error_app_frames,     type: :int,    default: 5
    option :source_lines_error_library_frames, type: :int,    default: 0
    option :source_lines_span_app_frames,      type: :int,    default: 5
    option :source_lines_span_library_frames,  type: :int,    default: 0
    option :span_frames_min_duration,          type: :float,  default: '5ms',   converter: Duration.new(default_unit: 'ms')
    option :stack_trace_limit,                 type: :int,    default: 999_999
    option :transaction_ignore_urls,           type: :list,   default: [],      converter: WildcardPatternList.new
    option :transaction_max_spans,             type: :int,    default: 500
    option :transaction_sample_rate,           type: :float,  default: 1.0,     converter: RoundFloat.new
    option :transaction_sample_rate_by_name,   type: :hash,   default: {},      converter: RoundFloatHashValue.new
    option :use_elastic_traceparent_header,    type: :bool,   default: true
    option :verify_server_cert,                type: :bool,   default: true

    def log_ecs_formatting
      log_ecs_reformatting
    end

    def log_ecs_formatting=(value)
      @options[:log_ecs_reformatting].set(value)
    end

    deprecate :log_ecs_formatting, :log_ecs_reformatting

    # rubocop:enable Layout/LineLength, Layout/ExtraSpacing
    def initialize(options = {})
      @options = load_schema

      assign(options)

      # Pick out config_file specifically as we need it now to load it,
      # but still need the other env vars to have precedence
      env = load_env
      if (env_config_file = env.delete(:config_file))
        self.config_file = env_config_file
      end

      assign(load_config_file)
      assign(env)

      yield self if block_given?

      if self.logger.nil? || self.log_path
        self.logger = build_logger
      end

      @__view_paths ||= []
      @__root_path ||= Dir.pwd
    end

    attr_accessor :__view_paths, :__root_path, :logger

    attr_reader :options

    def assign(update)
      return unless update
      update.each { |key, value| send(:"#{key}=", value) }
    end

    def available_instrumentations
      %w[
        action_dispatch
        azure_storage_table
        delayed_job
        dynamo_db
        elasticsearch
        faraday
        http
        json
        mongo
        net_http
        rake
        racecar
        redis
        resque
        s3
        sequel
        shoryuken
        sidekiq
        sinatra
        sneakers
        sns
        solid_queue
        sqs
        sucker_punch
        tilt
      ]
    end

    def enabled_instrumentations
      available_instrumentations - disable_instrumentations
    end

    def replace_options(new_options)
      return if new_options.nil? || new_options.empty?
      options_copy = @options.dup
      new_options.each do |key, value|
        options_copy.fetch(key.to_sym).set(value)
      end
      @options = options_copy
    end

    def app=(app)
      case app_type?(app)
      when :sinatra
        set_sinatra(app)
      when :rails
        set_rails(app)
      else
        self.service_name = 'ruby'
      end
    end

    def use_ssl?
      server_url.start_with?('https')
    end

    def collect_metrics?
      metrics_interval > 0
    end

    def span_frames_min_duration?
      span_frames_min_duration != 0
    end

    def span_frames_min_duration=(value)
      super
      @span_frames_min_duration_us = nil
    end

    def span_frames_min_duration_us
      @span_frames_min_duration_us ||= span_frames_min_duration * 1_000_000
    end

    def ssl_context
      return unless use_ssl?

      @ssl_context ||=
        OpenSSL::SSL::SSLContext.new.tap do |context|
          if server_ca_cert_file
            context.ca_file = server_ca_cert_file
          else
            context.cert_store =
              OpenSSL::X509::Store.new.tap(&:set_default_paths)
          end

          context.verify_mode =
            if verify_server_cert
              OpenSSL::SSL::VERIFY_PEER
            else
              OpenSSL::SSL::VERIFY_NONE
            end
        end
    end

    def inspect
      super.split.first + '>'
    end

    def version
      @version ||= ServerInfo.new(self).version
    end

    private

    def load_config_file
      return unless File.exist?(config_file)

      read = File.read(config_file)
      evaled = ERB.new(read).result
      YAML.safe_load(evaled)
    end

    def load_env
      @options.values.each_with_object({}) do |option, opts|
        next unless (value = ENV[option.env_key])
        opts[option.key] = value
      end
    end

    def build_logger
      if self.log_ecs_reformatting == 'override'
        begin
          return build_ecs_logger
        rescue LoadError
          logger.info "Attempted to use EcsLogging::Logger but the gem couldn't be " \
            "loaded so a ::Logger was created instead. Check if you have the `ecs-logging` " \
            "gem installed and attempt to start the agent again."
        end
      end

      Logger.new(log_path == '-' ? $stdout : log_path).tap do |logger|
        logger.level = log_level
      end
    end

    def build_ecs_logger
      require 'ecs_logging/logger'

      ::EcsLogging::Logger.new(log_path == '-' ? $stdout : log_path).tap do |logger|
        logger.level = log_level
      end
    end

    def app_type?(app)
      if defined?(::Rails::Application) && app.is_a?(::Rails::Application)
        return :rails
      end

      if app.is_a?(Class) && app.superclass.to_s == 'Sinatra::Base'
        return :sinatra
      end

      nil
    end

    def set_sinatra(app)
      self.service_name = format_name(service_name || app.to_s)
      self.framework_name = framework_name || 'Sinatra'
      self.framework_version = framework_version || ::Sinatra::VERSION
      self.__root_path = Dir.pwd
    end

    def set_rails(app)
      self.service_name ||= format_name(service_name || rails_app_name(app))
      self.framework_name ||= 'Ruby on Rails'
      self.framework_version ||= ::Rails::VERSION::STRING
      self.logger ||= ::Rails.logger
      self.log_level ||= ::Rails.logger.log_level

      self.__root_path = ::Rails.root.to_s
      self.__view_paths = app.config.paths['app/views'].existent +
                          [::Rails.root.to_s]
    end

    def rails_app_name(app)
      if ::Rails::VERSION::MAJOR >= 6
        app.class.module_parent_name
      else
        app.class.parent_name
      end
    end

    def format_name(str)
      str&.gsub('::', '_')
    end
  end
end
